File: examples/cpp/minimal-order-service/adapters/inbound/cli/order_controller.cpp
#include "adapters/inbound/cli/order_controller.hpp"
#include "application/ports/inbound/place_order_port.hpp"
#include "domain/order.hpp"
#include <stdexcept>

namespace adapters::inbound::cli {

OrderController::OrderController(
    std::shared_ptr<application::ports::inbound::PlaceOrderPort> placeOrderPort
) : placeOrderPort_(placeOrderPort) {
}

PlaceOrderResponse OrderController::placeOrder(const PlaceOrderRequest& request) {
    application::ports::inbound::PlaceOrderCommand command{
        request.amount,
        request.currency
    };
    
    auto order = placeOrderPort_->placeOrder(command);
    if (!order) {
        throw std::runtime_error("Failed to place order");
    }
    
    return PlaceOrderResponse{
        order->id(),
        order->total()->amount(),
        order->total()->currency(),
        order->status() == domain::OrderStatus::PAID ? "PAID" : "PENDING"
    };
}

} // namespace adapters::inbound::cli

File: examples/go/minimal-order-service/adapters/inbound/rest/order_controller.go
package rest

import (
	"github.com/example/order-service/application/ports/inbound"
	"github.com/example/order-service/domain"
)

// OrderController is an inbound adapter for REST API.
type OrderController struct {
	placeOrderPort inbound.PlaceOrderPort
}

// NewOrderController creates a new OrderController.
func NewOrderController(placeOrderPort inbound.PlaceOrderPort) *OrderController {
	return &OrderController{
		placeOrderPort: placeOrderPort,
	}
}

// PlaceOrderRequest represents the request DTO.
type PlaceOrderRequest struct {
	Amount   int64  `json:"amount"`
	Currency string `json:"currency"`
}

// PlaceOrderResponse represents the response DTO.
type PlaceOrderResponse struct {
	ID       string `json:"id"`
	Amount   int64  `json:"amount"`
	Currency string `json:"currency"`
	Status   string `json:"status"`
}

// PlaceOrder handles a place order request.
func (c *OrderController) PlaceOrder(req PlaceOrderRequest) (*PlaceOrderResponse, error) {
	command := inbound.PlaceOrderCommand{
		Amount:   req.Amount,
		Currency: req.Currency,
	}

	order, err := c.placeOrderPort.PlaceOrder(command)
	if err != nil {
		return nil, err
	}

	return &PlaceOrderResponse{
		ID:       order.ID(),
		Amount:   order.Total().Amount(),
		Currency: order.Total().Currency(),
		Status:   string(order.Status()),
	}, nil
}
